### ISSUE TITLE
Regression: referencing the `const`-defined variable can be done in a code before its declaration

### ISSUE THREAD

**Issue by mzgol** [jshint/jshint#1462]
Take the following code (from https://gist.github.com/mzgol/8305487):
```js
(function () {
  'use strict';
  const o = {
    f: function () {
      console.log(a);
    }
  };
  const a = 'aaa';
  o.f();
})();
```

In both IE11 & Chrome Canary with experimental JS flag on this code logs `'aaa'`. Referencing the `const`-defined variable in a code before its declaration is valid in ES6 as long as this code doesn't run before the declaration (I asked @rwaldron for confirmation).

jsHint used to not warn against this code when both `esnext` and `unused` were turned on (simply because `unused` didn't work for `const`-defined variables) but currently it errors on valid code, i.e. this is a regression.

**Comment by lukeapage** [comment]
@jugglinmike I think this is no longer an error on master and can be closed

**Comment by mzgol** [comment]
> @jugglinmike I think this is no longer an error on master and can be closed

I pasted this code into http://jshint.com/ with `/* jshint esnext: true */` added at the beginning and in the summary I see:

```
One unused variable
10	a
```

**Comment by lukeapage** [comment]
@mzgol sorry you are right, I mis intrepretted the issue - there are multiple const/let scope bug reports which overlap considerably.

Re-reading through it again I see it is not fixed on master, but is fixed in my PR #2344 and duplicated in #2292
